File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # abcabcbb。1， 2， 3， max(ans, 3) = 3, 
        charSet = set()
        l = 0
        ans = 0
        for r in range(len(s)):
            while s[r] in charSet:
                charSet.remove(s[l])
                l += 1
            ans = max(ans, r - l + 1)
            charSet.add(s[r])
        return ans

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int l = 0;
        int ans = 0;
        for (int r = 0; r < s.length(); r++) {
            while (set.contains(s.charAt(r))){
                set.remove(s.charAt(l));
                l++;
            }
            ans = Math.max(ans, r-l+1);
            set.add(s.charAt(r));
        }
        return ans;
    }   
}